blockprotocol/blockprotocol
# Improve codegen and expose entity label utility

## 🌟 What is the purpose of this PR?

Our codegen utility is a stand-in implementation which we intend to improve greatly in the coming weeks and months. There are many known shortcomings around its DX which are planned to be improved, but in lieu of those improvements the code it generates should at least be working (even if a bit unergonomic).

The current implementation fails on some strange edge-cases within types, including things like:
- Special characters in the type's `title` property
- Cyclical references between entity types across multiple link types which results in duplicate link definitions
- Clashing names between different classes of types (e.g. an entity type and property type with the same title)

This PR applies some _hotfixes_ to cover these edge-cases, but as is readily apparent, we should do a rewrite soon and make the code more maintainable and extensible.

Also this PR hoists the `parseLabelFromEntity` function and improves its implementation.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203358502199087/1204032127420832/f) _(internal)_
- [Asana task](https://app.asana.com/0/1203358502199087/1204080201525355/f) _(internal)_

## 🚫 Blocked by

N/A

## 🔍 What does this change?

Please see commits

## 📜 Does this require a change to the docs?

- To my knowledge, no

## ⚠️ Known issues

- The current deduplication logic doesn't capture all of the doc-comments
  <img width="933" alt="image" src="https://user-images.githubusercontent.com/25749103/221872956-dc225478-996e-4d61-9055-ded7c4175962.png">
- Type names can be pretty unwieldy and horrific

## 🐾 Next steps

- Rewrite the codegen

## 🛡 What tests cover this?

N/A

## ❓ How to test this?

1.  Checkout the branch
1.  Ensure `mock-block-dock` works as expected, look at the datastore graph visualization which uses `parseEntityFromLabel`
1. Try generating a complicated schema (see demo's below)

## 📹 Demo

### parseLabelFromEntity

#### Before

<img width="310" alt="image" src="https://user-images.githubusercontent.com/25749103/221874575-f7efb4a4-faf1-42e0-9c0d-98599b0fb72a.png">

#### After

<img width="310" alt="image" src="https://user-images.githubusercontent.com/25749103/221874675-442b6752-55a6-470b-ba71-5fac800e3f21.png">

### Codegen

#### Before

[Add pre-work demonstration](https://github.com/blockprotocol/blockprotocol/pull/1066/commits/6912ae8c9d0e3bbab1b1f0f55790a6fac9991ff5)

#### After

[Add post-work demonstration](https://github.com/blockprotocol/blockprotocol/pull/1066/commits/7d7cc151c55aeed9534e08ae526c7825ca2ced63)
